src/widget/city.c

```diff
@@ -293,41 +328,46 @@ static void handle_first_touch(map_tile *tile)
             }
         } else {
             if (first->has_started) {
-                if (data.selected_grid_offset != tile->grid_offset) {
+                if (data.selected_tile.grid_offset != tile->grid_offset) {
                     data.new_start_grid_offset = tile->grid_offset;
                 }
             }
             if (touch_not_click(first) && data.new_start_grid_offset) {
                 data.new_start_grid_offset = 0;
-                data.selected_grid_offset = 0;
+                data.selected_tile.grid_offset = 0;
                 building_construction_cancel();
                 build_start(tile);
             }
             build_move(tile);
-            if (data.selected_grid_offset != tile->grid_offset) {
-                data.selected_grid_offset = 0;
+            if (data.selected_tile.grid_offset != tile->grid_offset) {
+                data.selected_tile.grid_offset = 0;
             }
             if (first->has_ended) {
-                if (data.selected_grid_offset == tile->grid_offset) {
+                if (data.selected_tile.grid_offset == tile->grid_offset) {
                     build_end();
                     widget_city_clear_current_tile();
                     data.new_start_grid_offset = 0;
                 } else {
-                    data.selected_grid_offset = tile->grid_offset;
+                    data.selected_tile.grid_offset = tile->grid_offset;
                 }
             }
         }
         return;
     }
 
+    int size = building_properties_for_type(type)->size;
+    if (type == BUILDING_WAREHOUSE) {
+        size = 3;
+    }
+
     if (touch_was_click(first) && first->has_ended && data.capture_input &&
-        data.selected_grid_offset == tile->grid_offset) {
-        build_start(tile);
-        build_move(tile);
+        has_confirmed_construction(data.selected_tile.grid_offset, tile->grid_offset, size)) {
```

Comment:
How does this work for hippodromes and forts?

Reply:
Only the first section of each building (hippodrome and fort) can be clicked, clicking anywhere else moves the ghost. I know it's not perfect for those buildings, but it seemed a bit overkill to add the code for the entire ghost in those cases.

Reply:
Fair enough.